File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/state/operation/SeatState.java
package com.trtc.uikit.livekit.state.operation;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.trtc.tuikit.common.livedata.LiveData;
import com.trtc.tuikit.common.livedata.LiveMapData;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CopyOnWriteArrayList;

public class SeatState {
    private boolean                                  filterEmptySeat        = false;
    public  LiveData<List<SeatInfo>>                 seatList               =
            new LiveData<>(new CopyOnWriteArrayList<>());
    public  LiveData<LinkedHashSet<SeatApplication>> seatApplicationList    = new LiveData<>(new LinkedHashSet<>());
    public  LiveData<String>                         mySeatApplicationId    = new LiveData<>("");
    public  LiveMapData<String, SeatInvitation>      sentSeatInvitationMap  = new LiveMapData<>(new HashMap<>());
    public  LiveData<SeatInvitation>                 receivedSeatInvitation = new LiveData<>(new SeatInvitation());

    public void reset() {
        seatList.get().clear();
        seatApplicationList.get().clear();
        mySeatApplicationId.set("");
        sentSeatInvitationMap.get().clear();
        receivedSeatInvitation.get().reset();
    }

    public void setFilterEmptySeat(boolean filterEmptySeat) {
        this.filterEmptySeat = filterEmptySeat;
    }

    @NonNull
    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("{");
        builder.append("filterEmptySeat:").append(filterEmptySeat).append(",");
        for (SeatInfo seatInfo : seatList.get()) {
            builder.append("[");
            builder.append(seatInfo.index);
            builder.append(",");
            builder.append(seatInfo.userId.get());
            builder.append(",");
            builder.append(seatInfo.isLocked.get() ? "1" : "0");
            builder.append(",");
            builder.append(seatInfo.isAudioLocked.get() ? "1" : "0");
            builder.append("]");
        }
        builder.append("}");
        return builder.toString();
    }

    public void initSeatList(int maxSeatCount) {
        if (filterEmptySeat) {
            return;
        }
        List<SeatInfo> list = new ArrayList<>(maxSeatCount);
        for (int i = 0; i < maxSeatCount; i++) {
            SeatInfo seatInfo = new SeatInfo();
            seatInfo.index = i;
            list.add(seatInfo);
        }
        this.seatList.get().clear();
        this.seatList.addAll(list);
    }

    private void initSeatList(List<TUIRoomDefine.SeatInfo> list) {
        List<SeatInfo> newList = new ArrayList<>(list.size());
        for (TUIRoomDefine.SeatInfo info : list) {
            if (filterEmptySeat && TextUtils.isEmpty(info.userId)) {
                continue;
            }
            SeatInfo seatInfo = new SeatInfo();
            seatInfo.updateState(info);
            newList.add(seatInfo);
        }
        this.seatList.get().clear();
        this.seatList.addAll(newList);
    }

    public void updateSeatList(List<TUIRoomDefine.SeatInfo> list) {
        if (filterEmptySeat) {
            initSeatList(list);
            return;
        }
        if (list.size() != this.seatList.get().size()) {
            initSeatList(list);
            return;
        }
        for (int i = 0; i < list.size(); i++) {
            SeatInfo oldSeatInfo = this.seatList.get().get(i);
            TUIRoomDefine.SeatInfo newSeatInfo = list.get(i);
            if (oldSeatInfo == null || newSeatInfo == null) {
                continue;
            }
            oldSeatInfo.updateState(newSeatInfo);
        }
    }

    public void initSeatApplicationList(List<TUIRoomDefine.Request> list) {
        seatApplicationList.get().clear();
        List<SeatApplication> newList = new ArrayList<>();
        for (TUIRoomDefine.Request request : list) {
            SeatApplication application = new SeatApplication(request.requestId);
            application.updateState(request);
            newList.add(application);
        }
        seatApplicationList.addAll(newList);
    }

    public void addSeatApplication(TUIRoomDefine.Request request) {
        if (isRequestInvalid(request)) {
            return;
        }
        SeatApplication seatApplication = new SeatApplication(request.requestId);
        seatApplication.updateState(request);
        seatApplicationList.add(seatApplication);
    }

    public void removeSeatApplication(String id) {
        SeatApplication application = new SeatApplication(id);
        seatApplicationList.remove(application);
    }

    public void addSendSeatInvitation(TUIRoomDefine.Request request) {
        if (isRequestInvalid(request)) {
            return;
        }
        SeatInvitation seatInvitation = new SeatInvitation();
        seatInvitation.updateState(request);
        sentSeatInvitationMap.put(request.userId, seatInvitation);
    }

    public void removeSentSeatInvitation(String userId) {
        sentSeatInvitationMap.remove(userId);
    }

    public void addReceivedSeatInvitation(TUIRoomDefine.Request request) {
        if (isRequestInvalid(request)) {
            return;
        }
        SeatInvitation seatInvitation = new SeatInvitation();
        seatInvitation.updateState(request);
        receivedSeatInvitation.set(seatInvitation);
    }

    public void removeReceivedSeatInvitation() {
        SeatInvitation seatInvitation = receivedSeatInvitation.get();
        if (!TextUtils.isEmpty(seatInvitation.id)) {
            receivedSeatInvitation.get().reset();
            receivedSeatInvitation.notifyDataChanged();
        }
    }

    private boolean isRequestInvalid(TUIRoomDefine.Request request) {
        if (request == null) {
            return true;
        }
        if (TextUtils.isEmpty(request.userId)) {
            return true;
        }
        return TextUtils.isEmpty(request.requestId);
    }

    public static class SeatInfo {
        public int               index         = 0;
        public LiveData<String>  userId        = new LiveData<>("");
        public LiveData<String>  name          = new LiveData<>("");
        public LiveData<String>  avatarUrl     = new LiveData<>("");
        public LiveData<Boolean> isLocked      = new LiveData<>(false);
        public LiveData<Boolean> isAudioLocked = new LiveData<>(false);
        public LiveData<Boolean> isVideoLocked = new LiveData<>(false);

        public SeatInfo() {
        }

        public SeatInfo(String userId) {
            this.userId.set(userId);
        }

        @NonNull
        @Override
        public String toString() {
            return "SeatInfo{"
                    + "index=" + index
                    + ", userId=" + userId.get()
                    + ", isLocked=" + (isLocked.get() ? "1" : "0")
                    + ", isAudioLocked=" + (isAudioLocked.get() ? "1" : "0")
                    + ", isVideoLocked=" + (isVideoLocked.get() ? "1" : "0")
                    + '}';
        }

        @Override
        public boolean equals(@Nullable Object obj) {
            if (obj instanceof SeatInfo) {
                return this.userId.get().equals(((SeatInfo) obj).userId.get());
            }
            return false;
        }

        @Override
        public int hashCode() {
            return Objects.hash(userId.get());
        }

        public void updateState(TUIRoomDefine.SeatInfo seatInfo) {
            this.index = seatInfo.index;
            this.userId.set(seatInfo.userId, false);
            if (!TextUtils.isEmpty(seatInfo.userName)) {
                this.name.set(seatInfo.userName, false);
            } else {
                this.name.set(this.userId.get(), false);
            }
            this.avatarUrl.set(seatInfo.avatarUrl, false);
            this.isLocked.set(seatInfo.isLocked, false);
            this.isAudioLocked.set(seatInfo.isAudioLocked, false);
        }

        public void updateState(SeatInfo newSeatInfo) {
            this.index = newSeatInfo.index;
            this.userId.set(newSeatInfo.userId.get());
            if (!TextUtils.isEmpty(newSeatInfo.name.get())) {
                this.name.set(newSeatInfo.name.get(), false);
            } else {
                this.name.set(this.userId.get(), false);
            }
            this.avatarUrl.set(newSeatInfo.avatarUrl.get(), false);
            this.isLocked.set(newSeatInfo.isLocked.get(), false);
            this.isAudioLocked.set(newSeatInfo.isAudioLocked.get(), false);
        }
    }

    public static class SeatApplication {
        public String id        = "";
        public String userId    = "";
        public String userName  = "";
        public String avatarUrl = "";
        public long   timestamp;

        public SeatApplication(String id) {
            this.id = id;
        }

        public void updateState(TUIRoomDefine.Request request) {
            this.avatarUrl = request.avatarUrl;
            this.userId = request.userId;
            this.timestamp = request.timestamp;
            if (!TextUtils.isEmpty(request.userName)) {
                this.userName = request.userName;
            } else {
                this.userName = request.userId;
            }
        }

        @Override
        public boolean equals(@Nullable Object obj) {
            if (TextUtils.isEmpty(this.id)) {
                return false;
            }
            if (obj instanceof SeatApplication) {
                return this.id.equals(((SeatApplication) obj).id);
            }
            return false;
        }

        @Override
        public int hashCode() {
            return Objects.hash(id);
        }

        @NonNull
        @Override
        public String toString() {
            return "SeatApplication{"
                    + "id='" + id + '\''
                    + ", userId='" + userId + '\''
                    + ", timestamp=" + timestamp
                    + '}';
        }
    }

    public static class SeatInvitation {
        public String id        = "";
        public String userId    = "";
        public String userName  = "";
        public String avatarUrl = "";
        public long   timestamp;

        public void updateState(TUIRoomDefine.Request request) {
            this.id = request.requestId;
            this.avatarUrl = request.avatarUrl;
            this.userId = request.userId;
            this.timestamp = request.timestamp;
            if (!TextUtils.isEmpty(request.userName)) {
                this.userName = request.userName;
            } else {
                this.userName = request.userId;
            }
        }

        @NonNull
        @Override
        public String toString() {
            return "SeatInvitation{"
                    + "id='" + id + '\''
                    + ", userId='" + userId + '\''
                    + ", timestamp=" + timestamp
                    + '}';
        }

        public void reset() {
            this.id = "";
            this.userId = "";
            this.userName = "";
            this.avatarUrl = "";
            this.timestamp = 0;
        }
    }
}

File: Flutter/lib/state/operation/seat_state.dart
import 'package:flutter/material.dart';
import 'package:rtc_room_engine/api/room/tui_room_define.dart';

class SeatState {
  bool filterEmptySeat = false;
  final ValueNotifier<List<SeatInfo>> seatList = ValueNotifier([]);
  final ValueNotifier<List<SeatApplication>> seatApplicationList = ValueNotifier([]);
  final ValueNotifier<String> mySeatApplicationId = ValueNotifier('');

  void reset() {
    seatList.value.clear();
    seatApplicationList.value.clear();
    mySeatApplicationId.value = '';
  }

  void setFilterEmptySeat(bool filterEmptySeat) {
    this.filterEmptySeat = filterEmptySeat;
  }

  String toString() {
    StringBuffer buffer = StringBuffer('{');
    buffer.write("object");
    buffer.write('filterEmptySeat:');
    buffer.write(filterEmptySeat);
    buffer.write(',');
    for (SeatInfo seatInfo in seatList.value) {
      buffer.write('[');
      buffer.write(seatInfo.index);
      buffer.write(',');
      buffer.write(seatInfo.userId.value);
      buffer.write(',');
      buffer.write(seatInfo.isLocked.value);
      buffer.write(',');
      buffer.write(seatInfo.isAudioLocked.value);
      buffer.write(']');
    }
    buffer.write('}');
    return buffer.toString();
  }

  void updateSeatList(List<TUISeatInfo> list) {
    List<SeatInfo> newList = [];
    for (TUISeatInfo info in list) {
      if (filterEmptySeat && info.userId.isEmpty) {
        continue;
      }
      SeatInfo seatInfo = SeatInfo();
      seatInfo.updateState(info);
      newList.add(seatInfo);
    }
    seatList.value = newList;
  }

  void initSeatApplicationList(List<TUIRequest> list) {
    seatApplicationList.value.clear();
    List<SeatApplication> newList = [];
    for (TUIRequest request in list) {
      SeatApplication application = SeatApplication(request.requestId);
      application.updateState(request);
      newList.add(application);
    }
    seatApplicationList.value = newList;
  }

  void addSeatApplication(TUIRequest request) {
    if (request.requestId.isEmpty) {
      return;
    }
    if (request.requestAction != TUIRequestAction.requestToTakeSeat) {
      return;
    }
    SeatApplication seatApplication = SeatApplication(request.requestId);
    seatApplication.updateState(request);
    seatApplicationList.value.add(seatApplication);
    final applicationList = <SeatApplication>[];
    applicationList.addAll(seatApplicationList.value);
    seatApplicationList.value = applicationList;
  }

  void removeSeatApplication(String id) {
    SeatApplication application = SeatApplication(id);
    seatApplicationList.value.remove(application);
    final applicationList = <SeatApplication>[];
    applicationList.addAll(seatApplicationList.value);
    seatApplicationList.value = applicationList;
  }

  void removeSeatInfo(SeatInfo seatInfo) {
    seatList.value.remove(seatInfo);
    final list = <SeatInfo>[];
    list.addAll(seatList.value);
    seatList.value = list;
  }
}

class SeatInfo {
  int index = 0;
  ValueNotifier<String> userId = ValueNotifier('');
  ValueNotifier<String?> name = ValueNotifier('');
  ValueNotifier<String?> avatarUrl = ValueNotifier('');
  ValueNotifier<bool?> isLocked = ValueNotifier(false);
  ValueNotifier<bool?> isAudioLocked = ValueNotifier(false);
  ValueNotifier<bool?> isVideoLocked = ValueNotifier(false);

  SeatInfo() {}

  SeatInfo.fromUserId(String userId) {
    this.userId.value = userId;
  }

  @override
  String toString() {
    return 'SeatInfo{index=$index, userId=${userId.value}, isLocked=${isLocked.value}, '
        'isAudioLocked=${isAudioLocked.value}, isVideoLocked=${isVideoLocked.value}}';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is SeatInfo && other.userId.value == userId.value;
  }

  int get hashCode => userId.value.hashCode;

  void updateState(TUISeatInfo seatInfo) {
    index = seatInfo.index;
    userId.value = seatInfo.userId;
    if (seatInfo.userName != null && seatInfo.userName!.isNotEmpty) {
      name.value = seatInfo.userName;
    } else {
      name.value = userId.value;
    }
    avatarUrl.value = seatInfo.avatarUrl;
    isLocked.value = seatInfo.isLocked;
    isAudioLocked.value = seatInfo.isAudioLocked;
  }

  void updateNewState(SeatInfo newSeatInfo) {
    index = newSeatInfo.index;
    userId.value = newSeatInfo.userId.value;
    if (newSeatInfo.name.value != null && newSeatInfo.name.value!.isNotEmpty) {
      name.value = newSeatInfo.name.value;
    } else {
      name.value = userId.value;
    }
    avatarUrl.value = newSeatInfo.avatarUrl.value;
    isLocked.value = newSeatInfo.isLocked.value;
    isAudioLocked.value = newSeatInfo.isAudioLocked.value;
  }
}

class SeatApplication {
  String id = '';
  String userId = '';
  String? userName = '';
  String? avatarUrl = '';
  int timestamp = 0;

  SeatApplication(this.id);

  void updateState(TUIRequest request) {
    userName = request.userName;
    avatarUrl = request.avatarUrl;
    userId = request.userId;
    timestamp = request.timestamp;
  }

  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is SeatApplication && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() {
    return 'SeatApplication{id=\'$id\', userId=\'$userId\', timestamp=$timestamp}';
  }
}
